website/blog/2020-06-16-alluxio-datacaching.md

```diff
@@ -0,0 +1,243 @@
+3. If the secondary preferred worker is also busy, it assigns the split to the least busy worker.
+
+![](/img/blog/2020-06-16-alluxio-datacaching/presto_affinity_scheduler_algo.jpg)
+
+The definition of a busy node is defined by two configs: 
+1. Max splits per node:  *node-scheduler.max-splits-per-node*
+2. Max pending splits per task: *node-scheduler.max-pending-splits-per-task*
+
+Once the number of splits on one node exceeds one of the above configured limitations, this node would be treated as a busy node.
+<br><br>
+As it can be observed, node affinity is absolutely critical for cache effectiveness. Without node affinity, the same split may be processed by 
+different workers at different times, which can make caching the split data redundant. 
+<br><br>
+Due to this, if the affinity scheduler fails to assign the split to a preferred worker (because it was busy), it signals the assigned worker 
+to not cache the split data. It means the worker would only cache the split data if it is the primary or secondary preferred worker for the split. 
+
+## Alluxio data cache
+
+Alluxio file system is an open-source data orchestration system that is often used as a distributed caching service to Presto. 
+To achieve sub-second query latencies in our architecture, we want to further reduce the communication overhead between Presto and Alluxio. 
+As a result, core teams from Alluxio and Presto collaborated to carve out a single-node, embedded cache library from Alluxio service.
+<br><br>
+In particular, a Presto worker queries this Alluxio local cache inside the same JVM through a standard HDFS interface. 
+On a cache hit, Alluxio local cache directly reads data from the local disk and returns cache result to the Presto threads; 
+otherwise, it retrieves data from the remote data source, and caches the data on the local disk for followup queries. 
+This cache is completely transparent to Presto. In case the cache runs into issues (e.g., local disk failures), the Presto reads 
+fallback to the remote data source. This workflow is shown as the figure below.
+
+<br><br>
+![](/img/blog/2020-06-16-alluxio-datacaching/presto_alluxio_caching.jpg)
+<br><br>
+
+## Cache internals and configuration
+Our Alluxio data cache is a library residing in the Presto worker. It provides an HDFS-compatible interface “AlluxioCachingFileSystem” as the 
+main interface to Presto workers for all data access operations. 
+<br>
+There are some design choices under the hood:
+
+### Basic Caching Unit
+Both Alluxio experience and earlier experiments from the Facebook team suggested that reading, writing and evicting data in a fixed block size is 
+most efficient. Though in the Alluxio system, the default caching block size, 64MB, is fairly large mostly to reduce the storage and service pressure 
```

Comment:
In the Alluxio system the default caching block size is 64MB. This is fairly large mostly to reduce the storage and service pressure on the metadata service.

We significantly reduce the caching granularity because our our adaptation of the Alluxio data cache keep track of data and metadata locally. We default the cache granularity to units of 1MB "pages".